File: examples/webhooks/node/verify.ts
import { createHmac, timingSafeEqual } from 'crypto'

function safeEqual(a: string, b: string) {
  const aBuf = Buffer.from(a)
  const bBuf = Buffer.from(b)
  if (aBuf.length !== bBuf.length) return false
  return timingSafeEqual(aBuf, bBuf)
}

export function verifyWebhookV2(rawBody: string, headers: Record<string, string>, secret: string): boolean {
  const ts = headers['x-webhook-timestamp']
  const sigHeader = headers['x-webhook-signature-v2'] || ''
  const sig = sigHeader.replace(/^sha256=/, '')
  if (!ts || !sig) return false
  const now = Math.floor(Date.now() / 1000)
  if (Math.abs(now - Number(ts)) > 300) return false
  const base = `${ts}.${rawBody}`
  const expected = createHmac('sha256', secret).update(base).digest('hex')
  return safeEqual(expected, sig)
}

export function verifyWebhookLegacy(rawBody: string, headers: Record<string, string>, secret: string): boolean {
  const sigHeader = headers['x-webhook-signature'] || ''
  const sig = sigHeader.replace(/^sha256=/, '')
  if (!sig) return false
  const expected = createHmac('sha256', secret).update(rawBody).digest('hex')
  return safeEqual(expected, sig)
}

File: examples/webhooks/python/verify.py
import hmac
import hashlib
import time
from typing import Dict


def verify_webhook_v2(body: bytes, headers: Dict[str, str], secret: str) -> bool:
    ts = headers.get("x-webhook-timestamp")
    sigv2 = (headers.get("x-webhook-signature-v2") or "").replace("sha256=", "")
    if not ts or not sigv2:
        return False
    try:
        ts_int = int(ts)
    except Exception:
        return False
    # Replay window 5min
    if abs(int(time.time()) - ts_int) > 300:
        return False
    base = f"{ts}.{body.decode('utf-8')}".encode("utf-8")
    expected = hmac.new(secret.encode("utf-8"), base, hashlib.sha256).hexdigest()
    return hmac.compare_digest(expected, sigv2)


def verify_webhook_legacy(body: bytes, headers: Dict[str, str], secret: str) -> bool:
    sig = (headers.get("x-webhook-signature") or "").replace("sha256=", "")
    if not sig:
        return False
    expected = hmac.new(secret.encode("utf-8"), body, hashlib.sha256).hexdigest()
    return hmac.compare_digest(expected, sig)
